Satoshi notifications

lontivero opened the pull request:
# Push notifications for Satoshi identity using web sockets.

This PR is for the coordinator to send notifications about changes such as mining fee rates, exchange rates, new filters, round state changes, etc to all the clients. In this way clients don't need to request information using long pooling, instead the simple listen passively and get notifications about different events as soon as those occurs.

## Websocket infrastructure

The new server-side infrastructure components are mostly in the Backend's middleware folder. The solution is implemented as a ASP.NET middleware and it looks pretty much as a socket server implementation were after receiving a http request, if it is a protocol upgrade (websocket) request it accepts the connection, register the websocket in a list of websockets and start listening for messages from the connected client. The implementation is pretty generic because the idea is to have a reusable infra for implementing websocket communication for other identities too (Alice is the best candidate here)

## Satoshi messages

All what can be known by satoshi is implemented by the `SatoshiWebSocketHandler` handler. This sends mining fee rates, exchange rates and compact filters as soon as they are created. 

Upon establishing the connection, the client transmits the hash of the most recent known block to the coordinator, enabling the latter to start sending filters from that specific hash onwards. After all filters have been sent, it subscribes the socket to receive all new filters. 

----------

**Note:** Continuing with this PR requires lots of changes in many places and for that reason I think it will be better to start changing small things before even doing any further progress.

Fixes: https://github.com/zkSNACKs/WalletWasabi/issues/12117

Szpoti commented:
Tested on regtest, and there is a bug that we met on the review meeting as well:
When a transaction gets confirmed it doesn't get updated in the transaction history.
The pending icon is still visible, but in the Wallet Coins View, you can see them confirmed.
Restarting the wallet fixes this.

lontivero commented:
@turbolay @Szpoti any progress here?

Szpoti commented:
@turbolay created a test server where we can properly test this, so I'll do live testing and see if it all goes well.

lontivero commented on WalletWasabi/Services/SatoshiSynchronizer.cs:
```diff
@@ -0,0 +1,173 @@
+						break;
+
+					case ResponseMessage.HandshakeError:
+						await RewindAsync(144).ConfigureAwait(false);
```
The client connects to the server and sends the bestblockhash that it knew about last time it was connected. The server don't know about that block, obviously there was a reorg so, the client need to select a new starting synchronization point and minimize the failure rate. We use to assume 100 blocks everywhere even when a 100 blocks reorg is extremely unlikely, we could go back 6, 10, 31 or any other number of blocks. In this case 144 (one day) was chosen arbitrarily. Perhaps going back 10 block is more than enough.

lontivero replied:
If for some reason there is an error during the handshake then the server responses with `Handshake` error. However, there is only one valid reason covered by the protocol and that's the case when the block is not found. 

Could we send some extra information to the client like a reason? Yes, we could send BlockNotFound and InvalidRequest or similar. Currently, if you provide an invalid blockhash to the web controller it responses with a 400 BadRequest which is translated to a HttpRequestException what results in `BackendConnected = false`. That's sucks. So, if we get an InvalidRequest for example, that would mean we have a bug on the client.

In summary, i think it is not necessary.

lontivero commented on WalletWasabi.Backend/Middlewares/SatoshiWebSocketHandler.cs:
```diff
@@ -0,0 +1,213 @@
+
+			getFiltersChunk = GetFiltersBucketStartingFrom(lastTransmittedFilter);
+		}
+	}
```
Finding the block hash in the list of filters requires to traverse the whole list so, less you call the method less work for the CPU.

lontivero commented on WalletWasabi/Services/SatoshiSynchronizer.cs:
```diff
@@ -0,0 +1,173 @@
+						await _bitcoinStore.IndexStore.AddNewFiltersAsync([filter]).ConfigureAwait(false);
+						break;
+
+					case ResponseMessage.HandshakeError:
```
In regtest we remove all the filters when the backend starts, that means that a synchronized client knows more filters than the server because the server has to build them all again while the client already knows them all. For that reason the client tries to reconnect and once it does it, it send the bestknownblockchash to the server but the server still doesn't know about it.

This is something that will never happen in testnet or mainnet.

lontivero commented on WalletWasabi/Services/SatoshiSynchronizer.cs:
```diff
@@ -0,0 +1,173 @@
+			}
+		}
+
+		async Task RewindAsync(int count)
```
Good catch. I will take a look.

turbolay replied:
Waiting for #12778

lontivero commented on WalletWasabi/Services/SatoshiSynchronizer.cs:
```diff
@@ -0,0 +1,173 @@
+			}
+			catch (Exception e)
+			{
+				Logger.LogError(e);
```
Ohh I see, in case of cancellation it shouldn't log.

turbolay commented on WalletWasabi/Synchronizarion/FilterMessage.cs:
```diff
@@ -0,0 +1,33 @@
+	BlockHeight,
+	ExchangeRate,
+	MiningFeeRates
+}
```
I will live with it

turbolay commented:
I pushed some commits to fix the issues I found.
Commits are easy.
After #12778 merged, this PR should be ready to merged, I tested it again and didn't find additional issues

Szpoti commented:
I tested it 2 weeks ago when regtest server was set, didn't find any problems or errors, will do another round of testing with the new changes.

turbolay commented:
It has been decided to not have this for the release.

lontivero commented:
@turbolay I have rebased this branch on top of the released version and I have removed the merges and the commit fixing merging problems and so on. The idea is that the branch it already too big and I would like to keep it clean.

@turbolay @Szpoti  if this PR can be merged please approve it, otherwise let me know what is missing.

Szpoti commented on WalletWasabi/Synchronization/FilterMessage.cs:
```diff
@@ -0,0 +1,33 @@
+
+public enum RequestMessage
+{
+	BestKnowBlockHash
```
```suggestion
	BestKnownBlockHash
```

lontivero replied:
Done in [ba429eb](https://github.com/zkSNACKs/WalletWasabi/pull/12400/commits/ba429eba531e99fdfc8ccffef9b8ca016aca0d4a)

Szpoti reviewed (approved):
Just these small nits, otherwise tested and works.

lontivero commented on WalletWasabi.Backend/Middlewares/SatoshiWebSocketHandler.cs:
```diff
@@ -0,0 +1,213 @@
+		{
+			switch ((RequestMessage) buffer[0])
+			{
+				case RequestMessage.BestKnowBlockHash:
```
Done in [ba429eb](https://github.com/zkSNACKs/WalletWasabi/pull/12400/commits/ba429eba531e99fdfc8ccffef9b8ca016aca0d4a)
